fdbserver/ptxn/test/RealStorageServer.actor.cpp

```diff
@@ -0,0 +1,180 @@
+ * limitations under the License.
+ */
+
+#include "fdbrpc/simulator.h"
+#include "fdbserver/MockPeekCursor.h"
+#include "fdbserver/MockLogSystem.h"
+#include "fdbclient/StorageServerInterface.h"
+#include "fdbserver/IKeyValueStore.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "fdbserver/RecoveryState.h"
+#include "fdbserver/ptxn/test/Driver.h"
+
+#include "flow/UnitTest.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+
+namespace ptxn {
+
+struct ServerTestDriver {
+	ISimulator::ProcessInfo* previousProcess;
+	TaskPriority previousTask;
+
+	ACTOR static Future<Void> switchToServerProcess(ServerTestDriver* self) {
+		self->previousProcess = g_simulator.getCurrentProcess();
+		self->previousTask = g_pSimulator->getCurrentTask();
+		wait(g_simulator.onProcess(
+		    g_simulator.newProcess(
+		        "TestSystem",
+		        IPAddress(0x01010101),
+		        1,
+		        false,
+		        1,
+		        LocalityData(Optional<Standalone<StringRef>>(),
+		                     Standalone<StringRef>(deterministicRandom()->randomUniqueID().toString()),
+		                     Standalone<StringRef>(deterministicRandom()->randomUniqueID().toString()),
+		                     Optional<Standalone<StringRef>>()),
+		        ProcessClass(ProcessClass::TesterClass, ProcessClass::CommandLineSource),
+		        "",
+		        "",
+		        currentProtocolVersion),
+		    TaskPriority::DefaultYield));
+		Sim2FileSystem::newFileSystem();
+		FlowTransport::createInstance(false, 1);
+		return Void();
+	}
+
+	ACTOR static Future<Void> switchBack(ServerTestDriver* self) {
+		wait(g_simulator.onProcess(self->previousProcess, self->previousTask));
+		return Void();
+	}
+};
+
+struct StorageServerTestDriver : ServerTestDriver {
+	Reference<MockPeekCursor> mockPeekCursor;
+
+	// Default tag.
+	Tag tag = Tag(1, 1);
+	// Set by initEndpoints()
+	StorageServerInterface ssi;
+	ActorCollection actors = ActorCollection(false);
+
+	StorageServerTestDriver() : ServerTestDriver() { initEndpoints(); }
+
+	void initEndpoints() {
+		UID uid = nondeterministicRandom()->randomUniqueID();
+		ssi = StorageServerInterface(uid);
+		ssi.locality = LocalityData();
+		ssi.initEndpoints();
+	}
+
+	ACTOR static Future<Void> runStorageServer(StorageServerTestDriver* self) {
+		wait(delay(1));
+
+		StorageServerInterface& ssi = self->ssi;
+
+		KeyValueStoreType storeType = KeyValueStoreType::SSD_BTREE_V2;
+		state std::string folder = ".";
+		std::string fileName = joinPath(folder, "storage-" + ssi.id().toString() + "." + storeType.toString());
+		std::cout << "new Storage Server file name: " << fileName << std::endl;
+		deleteFile(fileName);
+		state IKeyValueStore* data = openKVStore(storeType, fileName, ssi.id(), 0);
+
+		state ReplyPromise<InitializeStorageReply> storageReady;
+
+		// Initialize dbInfo with mocked cursor and mocked loy system.
+
+		// TODO: Use Reference API?
+		// TODO: delete mockLogSystem eventually.
```

Comment:
Is there a reason this can't be a `unique_ptr`? I would strongly recommend that we do not commit code that contains known memory leaks.

Reply:
Using `unique_ptr` here will cause trouble because we are going to pass `mockLogSystem` as a void pointer. 

In `WorkerInterface.actor.h`, I cannot include `MockLogSystem` because it's a circular dependency (`MockLogSystem` -> `LogSystem` -> `WorkerInterface`). Forward declaration of `struct MockLogSystem` does not work either because ACTOR needs to know how to deref it:
```
struct MockLogSystem;
ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
                                 StorageServerInterface ssi,
                                 Tag seedTag,
                                 ReplyPromise<InitializeStorageReply> recruitReply,
                                 Reference<AsyncVar<ServerDBInfo>> db,
                                 std::string folder,
                                 Reference<MockLogSystem> mockLogSystem);
```

Reply:
Changing to `shared_ptr` works because its deleter is not selected at compile time so forward declaration is sufficient.